src/v/cloud_topics/level_zero/common/producer_queue.h

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "model/record.h"
+
+#include <memory>
+
+namespace cloud_topics::l0 {
+
+class producer_ticket;
+
+// A producer queue is a linearization point for producers within cloud topics.
+//
+// In cloud topics we accept requests in order to the batcher from the kafka
+// layer, and post upload we also need to release requests into raft in the same
+// order we recieved them. However we don't want to limit in flight requests to
+// one request per partition, we want to allow multiple per producer. So the
+// producer queue is the mechanism to enforce this ordering.
+//
+// Expected usage looks like this:
+//
+// ```
+// ss::promise<> enqueued_promise = ...;
+// ss::promise<> finished_promise = ...;
+// auto ticket = producer_queue.reserve(my_pid);
+// enqueued_promise.set_value();
+// co_await write_and_debounce(...);
+// co_await ticket.redeem(); // wait for previous replicate calls
+// auto stages = partition->replicate(...);
+// // the next request/ticket can be redeemed after the request is in raft
+// stages.request_enqueued.then([&ticket] { ticket.release(); });
+// stages.request_finished.forward_to(finished_promise);
+// ```
+//
+// NOTE: This class' memory usage is bounded by the concurrency of in flight
+// requests and should scale ~linearly with the number of concurrent produce
+// requests.
+class producer_queue {
+public:
+    producer_queue();
+    ~producer_queue();
+    producer_queue(const producer_queue&) = delete;
+    producer_queue(producer_queue&&) noexcept;
+    producer_queue& operator=(const producer_queue&) = delete;
+    producer_queue& operator=(producer_queue&&) noexcept;
+
+    // Reserve a position within the queue for a given producer.
+    // The ticket can be redeemed later on to wait for previous tickets to be
+    // released. This creates a queuing effect per producer id.
+    //
+    // NOTE: If the producer id is model::no_producer_id, then this ticket
+    // is basically a noop and doesn't wait or preserve any ordering.
+    producer_ticket reserve(model::producer_id);
```

Comment:
maybe this should be futurized so we could apply additional limits per producer id / partition?

Reply:
I'll make that change when it's needed. I'm up for additional limits, but I'd like to find the need for that first. In reality this method is really only used in a single place so it's not much to do that.